### PR TITLE
Add support for custom rules that use references

### PR DESCRIPTION
**Pull request by ChrisExP**
Laravel allows creating custom validation rules. These could reference
other fields sent in the request. Due to the nature of graphQL when you
generate the schema the input types could be placed in multiple contexts
so the absolute field path cannot be known when generating the rules.
Lighthouse deals with the default validation fields by adding the
argument path to the parameters that need it but it didn't support
custom validation.

- [x] Added or updated tests
- [x] Documented user facing changes
- [x] Updated CHANGELOG.md

Pull request for #1901

**Changes**

This change allows the developer to wrap their custom rule in a
`with_reference` rule specifying the rule and the index of the parameter
that should be used as a reference. Alternatively with custom rule
classes the rule can implement the `WithReferenceRule` interface and
manipulate the path with the `setArgumentPath` method.

### CODE REVIEW

**Comment by ChrisExP** [comment]
I was able to get the composition working without even creating a rule or an extension because the rule has to go through the rule gatherer however it is defined, I was able to just unwrap it there.

However I couldn't think of a way to use composition for custom rule classes because it is up to the developer how that accesses the other fields. There was no way I could think to get around that, so I fell back to the interface method I suggested.

I am happy to make any changes if you have other suggestions.

I wasn't sure if this was a big enough change to add it into the documentation.

File: src/Support/Contracts/WithReferenceRule.php [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<?php
+
+namespace Nuwave\Lighthouse\Support\Contracts;
+
+interface WithReferenceRule
+{
+    public function setArgumentPath(string $argumentPath): void;
```
Comment (spawnia) [inline src/Support/Contracts/WithReferenceRule.php:7]: This should have a docblock explaining what the argument path exactly is and how to use it.

File: src/Validation/RulesGatherer.php [hunk 2]
```diff
@@ -224,6 +225,10 @@ protected function qualifyArgumentReferences(array $rules, array $argumentPath):
         return array_map(
             static function ($rule) use ($argumentPath) {
                 if (is_object($rule)) {
+                    if ($rule instanceof WithReferenceRule) {
+                        $rule->setArgumentPath(implode('.', $argumentPath));
```
Comment (spawnia) [inline src/Validation/RulesGatherer.php:229]: We can simply pass the array, access to more structured data is generally better.

File: src/Validation/RulesGatherer.php [hunk 3]
```diff
@@ -238,6 +243,20 @@ static function ($rule) use ($argumentPath) {
                 $name = $parsed[0];
                 $args = $parsed[1];
 
+                if ($name === 'WithReference') {
+                    $indexes = explode('_', $args[1]);
+                    array_splice($args, 1, 1);
+                    foreach ($indexes as $index) {
+                        $index = (int) $index + 1;
```
Comment (spawnia) [inline src/Validation/RulesGatherer.php:250]:
```suggestion
                        // Skipping over the first index, which is the name
                        $index = (int) $index + 1;
```

**Comment by spawnia** [comment]
This is a new feature and thus warrants a minor version bump and documentation. How else is anyone supposed to know about this?

**Comment by ChrisExP** [comment]
The tests are failing because in Laravel 8 they added an interface that allows custom rule classes to interact with the validator data.
Before that this wouldn't have been possible with custom rule classes so the test wouldn't apply to versions before 8.
Is there any way to tell it to skip a test and for PHPStan to skip a file for Laravel versions below 8, or should I leave the test out?

**Comment by spawnia** [comment]
You can just test this without the interface. The validation rule for the test does not have to make sense, for our purposes it is probably easiest if it always fails and includes the argument path in the resulting error message.

**Pull request merged**